File: cane-csharp/cane-core/Runtime/Animation/Playback.cs
using System;
using System.Collections.Generic;

namespace Cane
{
    public enum TrackBlend { Replace, Additive }
    public sealed class TrackOptions
    {
        public float? Alpha { get; set; }
        public float? TimeScale { get; set; }
        public bool? Looping { get; set; }
        public TrackBlend? Blend { get; set; }
        public float? EventThreshold { get; set; }
        public float? AttachmentThreshold { get; set; }
        public float? DrawOrderThreshold { get; set; }
        public bool? HoldPrevious { get; set; }
    }
    public sealed class QueuedEntryOptions
    {
        public float? DelaySeconds { get; set; }
        public float? MixDurationSeconds { get; set; }
        public bool? Looping { get; set; }
    }
    public readonly struct SamplingOptions
    {
        public readonly bool Stepped;
        public readonly float FramesPerSecond;
        public readonly float? FrameStepSeconds;
        public SamplingOptions(bool stepped = false, float framesPerSecond = 0)
        { Stepped = stepped; FramesPerSecond = framesPerSecond; FrameStepSeconds = null; }
        private SamplingOptions(float frameStepSeconds, bool stepped)
        { Stepped = stepped; FramesPerSecond = 0; FrameStepSeconds = frameStepSeconds; }
        public static SamplingOptions FixedFrame(float frameStepSeconds, bool stepped = false) => new SamplingOptions(frameStepSeconds, stepped);
        internal float Time(float value)
        {
            if (!FrameStepSeconds.HasValue && FramesPerSecond == 0) return value;
            float step = FrameStepSeconds ?? (1 / FramesPerSecond);
            float result = (float)(Math.Floor((double)value / step + 0.5) * step);
            if (!Numeric.Finite(result)) throw new RuntimeException(RuntimeErrorCode.NonFinite, "apply", "Fixed-frame sample time became non-finite.", "sampling");
            return result;
        }
        internal void Validate(string op)
        {
            Numeric.RequireFinite(FramesPerSecond, op, "sampling.fps");
            if (FramesPerSecond < 0) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Sampling fps must be positive or zero for authored sampling.", "sampling.fps");
            if (FrameStepSeconds.HasValue || FramesPerSecond > 0)
            {
                float step = FrameStepSeconds ?? (1 / FramesPerSecond); Numeric.RequireFinite(step, op, "sampling.frameStepSeconds");
                if (step <= 0) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Frame step must be positive.", "sampling.frameStepSeconds");
            }
        }
    }
    public sealed class RuntimeStep
    {
        public bool Changed { get; }
        public float TimeSeconds { get; }
        public ulong FrameSequence { get; }
        public IReadOnlyList<RuntimeEvent> Events { get; }
        internal RuntimeStep(bool changed, RuntimeFrame frame, RuntimeEvent[] events)
        { Changed = changed; TimeSeconds = frame.TimeSeconds; FrameSequence = frame.Sequence; Events = Array.AsReadOnly(events); }
    }
    public sealed class TrackState
    {
        public int TrackIndex { get; }
        public string? AnimationId { get; }
        public float TrackTime { get; }
        public float AnimationTime { get; }
        public float AnimationStart { get; }
        public float AnimationEnd { get; }
        public float DelaySeconds { get; }
        public float? TrackEnd { get; }
        public float Alpha { get; }
        public float TimeScale { get; }
        public bool Looping { get; }
        public TrackBlend Blend { get; }
        public float MixTime { get; }
        public float MixDuration { get; }
        public float MixProgress { get; }
        public float EventThreshold { get; }
        public float AttachmentThreshold { get; }
        public float DrawOrderThreshold { get; }
        public bool HoldPrevious { get; }
        public int QueuedEntryCount { get; }
        internal TrackState(int index, Animation.Entry e, int count)
        {
            TrackIndex = index; AnimationId = e.Clip?.Id; TrackTime = e.Time; AnimationTime = e.PoseTime;
            AnimationStart = e.Start; AnimationEnd = e.End; DelaySeconds = e.Delay; TrackEnd = e.TrackEnd; Alpha = e.Alpha; TimeScale = e.Rate;
            Looping = e.Loop; Blend = e.Additive ? TrackBlend.Additive : TrackBlend.Replace; MixTime = e.MixTime;
            MixDuration = e.MixDuration; MixProgress = e.Progress; EventThreshold = e.EventThreshold;
            AttachmentThreshold = e.AttachmentThreshold; DrawOrderThreshold = e.DrawOrderThreshold;
            HoldPrevious = e.Hold; QueuedEntryCount = count;
        }
    }
    public sealed class QueuedEntry
    {
        public int TrackIndex { get; }
        public int QueueIndex { get; }
        public string? AnimationId { get; }
        public float DelaySeconds { get; }
        public float MixDuration { get; }
        public bool Looping { get; }
        internal QueuedEntry(int track, int i, Animation.Entry e)
        { TrackIndex = track; QueueIndex = i; AnimationId = e.Clip?.Id; DelaySeconds = e.Delay; MixDuration = e.MixDuration; Looping = e.Loop; }
    }
}

namespace Cane.Animation
{
    internal sealed class Entry
    {
        internal Clip? Clip;
        internal float Time, Start, End, Delay, MixTime, MixDuration, Alpha = 1, Rate = 1;
        internal float EventThreshold, AttachmentThreshold, DrawOrderThreshold;
        internal float? TrackEnd;
        internal bool Loop, Additive, Hold, CursorInitialized;
        internal Entry? From;
        internal long EventOrder;
        internal float Duration => Math.Max(0, End - Start);
        internal float Progress => From == null || MixDuration == 0 ? 1 : Numeric.Clamp(MixTime / MixDuration, 0, 1);
        internal float PoseTime
        {
            get { float d = Duration; if (d == 0) return Start; float t = Loop ? Time % d : Numeric.Clamp(Time, 0, d); if (t < 0) t += d; return Start + t; }
        }
        internal Entry(Clip? clip, bool loop, float mix)
        { Clip = clip; Loop = loop; End = clip?.Duration ?? 0; MixDuration = mix; }
        internal Entry Clone() { var next = (Entry)MemberwiseClone(); next.From = From?.Clone(); return next; }
    }
    internal sealed class Track
    {
        internal Entry Current;
        internal readonly List<Entry> Queue = new List<Entry>();
        internal Track(Entry entry) { Current = entry; }
        internal Track Clone() { var copy = new Track(Current.Clone()); foreach (Entry e in Queue) copy.Queue.Add(e.Clone()); return copy; }
    }
    internal sealed class Playback
    {
        internal readonly SortedDictionary<int, Track> Tracks = new SortedDictionary<int, Track>();
        internal readonly Dictionary<(string, string), float> Mixes = new Dictionary<(string, string), float>();
        internal float DefaultMix, DetachedTime;
        internal int AllocatedTracks;
        internal long NextEventOrder;
        internal Constraints.PhysicsStore Physics = new Constraints.PhysicsStore();
        internal Playback Clone()
        {
            var copy = new Playback { DefaultMix = DefaultMix, DetachedTime = DetachedTime, AllocatedTracks = AllocatedTracks, NextEventOrder = NextEventOrder, Physics = Physics.Clone() };
            foreach (var p in Tracks) copy.Tracks.Add(p.Key, p.Value.Clone());
            foreach (var p in Mixes) copy.Mixes.Add(p.Key, p.Value); return copy;
        }
        internal float Time => Tracks.TryGetValue(0, out Track? t) ? t.Current.PoseTime : DetachedTime;
        internal float Mix(Entry? from, Clip? to) => from?.Clip == null || to == null ? DefaultMix : Mixes.TryGetValue((from.Clip.Id, to.Id), out float m) ? m : DefaultMix;
        internal void Replace(int index, Clip? clip, bool loop, float? mix, EventSink events)
        {
            AllocatedTracks = Math.Max(AllocatedTracks, index + 1);
            Tracks.TryGetValue(index, out Track? old); float duration = old == null ? 0 : mix ?? Mix(old.Current, clip);
            var entry = new Entry(clip, loop, duration) { EventOrder = NextEventOrder++ };
            if (old != null)
            {
                events.Lifecycle("interrupt", index, old.Current, 0);
                if (duration == 0) EndChain(old.Current, index, 0, events); else entry.From = old.Current;
                foreach (var queued in old.Queue) events.Lifecycle("dispose", index, queued, 0);
            }
            Tracks[index] = new Track(entry); events.Lifecycle("start", index, entry, 0);
        }
        internal void Enqueue(int index, Clip? clip, bool loop, float delay, float? mix, EventSink events)
        {
            if (!Tracks.TryGetValue(index, out Track? track)) { Replace(index, clip, loop, 0, events); return; }
            Entry previous = track.Queue.Count == 0 ? track.Current : track.Queue[track.Queue.Count - 1];
            float duration = mix ?? Mix(previous, clip);
            track.Queue.Add(new Entry(clip, loop, duration) { EventOrder = NextEventOrder++, Delay = delay > 0 ? delay : Math.Max(0, previous.Duration - duration + delay) });
        }
        internal void Clear(int index, EventSink events, double offset = 0)
        {
            if (!Tracks.TryGetValue(index, out Track? t)) return;
            EndChain(t.Current, index, offset, events); foreach (var e in t.Queue) events.Lifecycle("dispose", index, e, offset); Tracks.Remove(index);
        }
        internal static void EndChain(Entry? e, int track, double offset, EventSink events)
        { long order = e == null ? 0 : e.EventOrder * 2 + 1; while (e != null) { events.Lifecycle("end", track, e, offset, order); events.Lifecycle("dispose", track, e, offset, order); e = e.From; } }
        internal void Apply(Pose pose, SamplingOptions sampling)
        {
            var lower = pose.LowerTrackProperties; lower.Clear();
            foreach (var track in Tracks.Values)
            {
                ApplyChain(pose, track.Current, null, 1, lower, sampling);
                for (Entry? e = track.Current; e != null; e = e.From) if (e.Clip != null) lower.UnionWith(e.Clip.Properties);
            }
        }
        private static void ApplyChain(Pose pose, Entry e, Entry? incoming, float fade, HashSet<string> lower, SamplingOptions sampling)
        {
            float progress = e.Progress;
            if (e.From != null) ApplyChain(pose, e.From, e, e.Clip == null && !e.Hold ? fade * (1 - progress) : fade, lower, sampling);
            if (e.Clip == null) return;
            float alpha = fade * e.Alpha * progress;
            var layer = new Layer { Clip = e.Clip, Time = sampling.Time(e.PoseTime), Alpha = alpha, Additive = e.Additive, Stepped = sampling.Stepped };
            if (incoming != null)
            {
                float mix = incoming.Progress;
                layer.Attachments = mix < e.AttachmentThreshold; layer.DrawOrder = mix < e.DrawOrderThreshold;
                if (incoming.Clip != null && !incoming.Hold)
                    layer.PropertyAlpha = property => incoming.Clip.Properties.Contains(property) && !lower.Contains(property) ? alpha : alpha * (1 - mix);
            }
            Sampling.Apply(pose, layer);
        }
        internal void Update(float delta, EventSink events)
        {
            DetachedTime += delta; if (!Numeric.Finite(DetachedTime)) throw new RuntimeException(RuntimeErrorCode.NonFinite, events.Operation, "Playback clock overflowed.", "deltaSeconds");
            var indices = new List<int>(Tracks.Keys);
            foreach (int index in indices)
            {
                double offset = 0, remaining = delta;
                while (Tracks.TryGetValue(index, out Track? track))
                {
                    Entry e = track.Current; double span = remaining;
                    bool promote = track.Queue.Count != 0 && e.Rate > 0;
                    double queueAt = promote ? Math.Max(0, ((double)track.Queue[0].Delay - e.Time) / e.Rate) : double.PositiveInfinity;
                    double endAt = e.TrackEnd.HasValue && e.Rate > 0 ? Math.Max(0, ((double)e.TrackEnd.Value - e.Time) / e.Rate) : double.PositiveInfinity;
                    double mixAt = NextMix(e);
                    span = Math.Min(span, Math.Min(queueAt, Math.Min(endAt, mixAt)));
                    AdvanceChain(e, null, index, offset, span, events);
                    offset += span; remaining = Math.Max(0, remaining - span);
                    bool boundary = false;
                    if (queueAt <= span && queueAt <= endAt)
                    {
                        Entry next = track.Queue[0]; track.Queue.RemoveAt(0);
                        events.Lifecycle("interrupt", index, e, offset);
                        if (next.MixDuration == 0) EndChain(e, index, offset, events); else next.From = e;
                        track.Current = next; events.Lifecycle("start", index, next, offset); boundary = true;
                    }
                    else if (endAt <= span) { Clear(index, events, offset); break; }
                    else boundary = CompleteMixes(e, index, offset, events);
                    if (remaining == 0 && !boundary) break;
                    if (span == 0 && !boundary) break;
                }
            }
        }
        private static double NextMix(Entry e)
        {
            double at = double.PositiveInfinity;
            for (Entry? entry = e; entry?.From != null; entry = entry.From) at = Math.Min(at, Math.Max(0, (double)entry.MixDuration - entry.MixTime));
            return at;
        }
        private static bool CompleteMixes(Entry e, int index, double offset, EventSink events)
        {
            if (e.From == null) return false;
            if (e.MixDuration == 0 || e.MixTime >= e.MixDuration)
            { EndChain(e.From, index, offset, events); e.From = null; return true; }
            return CompleteMixes(e.From, index, offset, events);
        }
        private static void AdvanceChain(Entry e, Entry? incoming, int index, double offset, double delta, EventSink events)
        {
            if (e.From != null) AdvanceChain(e.From, e, index, offset, delta, events);
            float previous = e.Time, next = (float)((double)previous + (float)delta * e.Rate);
            if (!e.Loop) next = Math.Max(0, next);
            if (!Numeric.Finite(next)) throw new RuntimeException(RuntimeErrorCode.NonFinite, events.Operation, "Track clock overflowed.", "timeSeconds");
            EventCrossings.Collect(e, incoming, index, previous, next, offset, delta, events);
            e.Time = next; e.MixTime += (float)delta;
            if (!Numeric.Finite(e.MixTime)) throw new RuntimeException(RuntimeErrorCode.NonFinite, events.Operation, "Mix clock overflowed.", "mixTime");
        }
    }
}

File: cane-cpp/cane-core/src/animation/playback.cpp
#include "playback.hpp"
#include "runtime_data_internal.hpp"
#include <limits>

namespace cane::animation {
namespace {
void index_valid(std::uint32_t index, const char* op) {
    if (index > 4095) throw Error(ErrorCode::invalid_argument, op, "Track index must be in 0..4095.", "trackIndex");
}
void finite(float value, const char* op, const char* field) {
    if (!std::isfinite(value)) throw Error(ErrorCode::non_finite, op, "Playback scalar must be finite.", field);
}
void nonnegative(float value, const char* op, const char* field) {
    finite(value, op, field);
    if (value < 0) throw Error(ErrorCode::invalid_argument, op, "Playback scalar must be non-negative.", field);
}
void unit(const std::optional<float>& value, const char* op, const char* field) {
    if (!value) return;
    finite(*value, op, field);
    if (*value < 0 || *value > 1) throw Error(ErrorCode::invalid_argument, op, "Playback weight must be in 0..1.", field);
}
float progress(const Track& track, std::size_t at) {
    const auto& e = track.chain[at];
    return at == 0 || e.mix_duration == 0 ? 1 : std::clamp(e.mix_time / e.mix_duration, 0.0f, 1.0f);
}
}
float TrackEntry::pose_time() const noexcept {
    const float d = duration(); if (d == 0) return start;
    float value = looping ? std::fmod(time, d) : std::clamp(time, 0.0f, d);
    if (value < 0) value += d;
    return start + value;
}
float Playback::time() const noexcept {
    const auto found = tracks_.find(0); return found == tracks_.end() ? detached_time_ : found->second.chain.back().pose_time();
}
Playback Playback::clone_configuration() const {
    Playback result(data_); result.default_mix_ = default_mix_; result.mixes_ = mixes_; return result;
}
Playback Playback::rebind(RuntimeData data, const char* operation) const {
    auto result = *this; result.data_ = std::move(data);
    const auto bind = [&](TrackEntry& entry) {
        if (!entry.clip) return;
        entry.clip = result.resolve(entry.clip->id, operation);
        if (entry.start < 0 || entry.end < entry.start || entry.end > entry.clip->duration)
            throw Error(ErrorCode::invalid_argument, operation, "Active or queued animation range no longer fits the clip.", "animationRange", entry.clip->id);
    };
    for (auto& track : result.tracks_) {
        for (auto& entry : track.second.chain) bind(entry);
        for (auto& entry : track.second.queue) bind(entry);
    }
    for (const auto& pair : result.mixes_) { (void)result.resolve(pair.first.first, operation); (void)result.resolve(pair.first.second, operation); }
    return result;
}
void Playback::set_detached_time(float time) { nonnegative(time, "setAnimationTime", "timeSeconds"); detached_time_ = time; }
Playback Playback::reconcile(RuntimeData data) const {
    auto result = *this; result.data_ = std::move(data);
    const auto& next = detail::RuntimeDataAccess::get(result.data_);
    const auto bind = [&](TrackEntry& entry) {
        if (!entry.clip) return true;
        const auto at = next.index.catalogs[10].find(entry.clip->id);
        if (at == next.index.catalogs[10].end()) return false;
        const bool full_range = entry.start == 0 && entry.end == entry.clip->duration;
        entry.clip = &next.clips[at->second];
        if (full_range) { entry.start = 0; entry.end = entry.clip->duration; }
        else { entry.start = std::min(entry.start, entry.clip->duration); entry.end = std::min(entry.end, entry.clip->duration); }
        // Absolute elapsed time, event identity/cursor and mix progress remain unchanged.
        return true;
    };
    for (auto track = result.tracks_.begin(); track != result.tracks_.end();) {
        auto& chain = track->second.chain; std::size_t drop = 0;
        for (std::size_t i = 0; i < chain.size(); ++i) if (!bind(chain[i])) drop = i + 1;
        if (drop == chain.size()) { track = result.tracks_.erase(track); continue; }
        chain.erase(chain.begin(), chain.begin() + static_cast<std::ptrdiff_t>(drop));
        auto& queue = track->second.queue;
        queue.erase(std::remove_if(queue.begin(), queue.end(), [&](auto& e) { return !bind(e); }), queue.end());
        ++track;
    }
    for (auto pair = result.mixes_.begin(); pair != result.mixes_.end();) {
        if (!next.index.catalogs[10].count(pair->first.first) || !next.index.catalogs[10].count(pair->first.second)) pair = result.mixes_.erase(pair);
        else ++pair;
    }
    return result;
}
const Clip* Playback::resolve(std::optional<std::string_view> id, const char* operation) const {
    if (!id) return nullptr;
    const auto& state = detail::RuntimeDataAccess::get(data_); const auto found = state.index.catalogs[10].find(std::string(*id));
    if (found == state.index.catalogs[10].end()) throw Error(ErrorCode::not_found, operation, "Animation does not exist.", "animationId", std::string(*id));
    return &state.clips[found->second];
}
Track& Playback::require_track(std::uint32_t index, const char* operation) {
    index_valid(index, operation); const auto found = tracks_.find(index);
    if (found == tracks_.end()) throw Error(ErrorCode::not_found, operation, "Current track entry does not exist.", "trackIndex");
    return found->second;
}
TrackEntry Playback::make_entry(const Clip* clip, bool looping, float mix, const char* operation) {
    if (next_identity_ == std::numeric_limits<std::uint64_t>::max()) throw Error(ErrorCode::resource_limit, operation, "Playback entry identities are exhausted.", "trackIndex");
    TrackEntry e; e.clip = clip; e.end = clip ? clip->duration : 0; e.looping = looping;
    e.mix_duration = mix; e.identity = next_identity_++; return e;
}
float Playback::mix(const TrackEntry& from, const Clip* to) const {
    if (!from.clip || !to) return default_mix_;
    const auto found = mixes_.find({from.clip->id, to->id}); return found == mixes_.end() ? default_mix_ : found->second;
}
void Playback::set_default_mix(float seconds) { nonnegative(seconds, "setDefaultMix", "durationSeconds"); default_mix_ = seconds; }
void Playback::set_mix(std::string_view from, std::string_view to, float seconds) {
    nonnegative(seconds, "setMix", "durationSeconds"); (void)resolve(from, "setMix"); (void)resolve(to, "setMix");
    mixes_[{std::string(from), std::string(to)}] = seconds;
}
void Playback::retire(const Track& track, std::size_t count, std::uint32_t index, double offset, EventSink& sink) {
    if (count == 0) return;
    const auto block_order = track.chain[count - 1].identity;
    for (std::size_t at = count; at-- > 0;) {
        const auto view = track.chain[at].event_view();
        sink.lifecycle(EventKind::end, index, view, offset, block_order);
        sink.lifecycle(EventKind::dispose, index, view, offset, block_order);
    }
}
void Playback::replace(std::uint32_t index, std::optional<std::string_view> id, bool looping, std::optional<float> seconds, EventSink& sink) {
    const auto* op = sink.operation().c_str(); index_valid(index, op); if (seconds) nonnegative(*seconds, op, "mixSeconds");
    const auto* clip = resolve(id, op); const auto found = tracks_.find(index);
    const float duration = found == tracks_.end() ? 0 : seconds.value_or(mix(found->second.chain.back(), clip));
    auto next = make_entry(clip, looping, duration, op);
    if (found != tracks_.end()) {
        auto& old = found->second; sink.lifecycle(EventKind::interrupt, index, old.chain.back().event_view(), 0);
        if (duration == 0) retire(old, old.chain.size(), index, 0, sink);
        for (const auto& queued : old.queue) sink.lifecycle(EventKind::dispose, index, queued.event_view(), 0);
        if (duration == 0) old.chain.clear();
        old.queue.clear(); old.chain.push_back(std::move(next));
    } else { Track track; track.chain.push_back(std::move(next)); tracks_.emplace(index, std::move(track)); }
    allocated_tracks_ = std::max(allocated_tracks_, index + 1);
    sink.lifecycle(EventKind::start, index, tracks_.at(index).chain.back().event_view(), 0);
}
void Playback::enqueue(std::uint32_t index, std::optional<std::string_view> id, bool looping, float delay,
                        std::optional<float> seconds, EventSink& sink) {
    const auto* op = sink.operation().c_str(); index_valid(index, op); finite(delay, op, "delaySeconds");
    if (seconds) nonnegative(*seconds, op, "mixSeconds");
    const auto* clip = resolve(id, op);
    const auto found = tracks_.find(index);
    if (found == tracks_.end()) { replace(index, id, looping, 0.0f, sink); return; }
    auto& track = found->second; const auto& previous = track.queue.empty() ? track.chain.back() : track.queue.back();
    const float duration = seconds.value_or(mix(previous, clip));
    const float threshold = delay > 0 ? delay : std::max(0.0f, previous.duration() - duration + delay);
    finite(threshold, op, "delaySeconds"); auto e = make_entry(clip, looping, duration, op); e.delay = threshold;
    track.queue.push_back(std::move(e));
}
void Playback::clear(std::uint32_t index, EventSink& sink, double offset) {
    index_valid(index, sink.operation().c_str()); const auto found = tracks_.find(index); if (found == tracks_.end()) return;
    const auto& track = found->second; retire(track, track.chain.size(), index, offset, sink);
    for (const auto& queued : track.queue) sink.lifecycle(EventKind::dispose, index, queued.event_view(), offset);
    tracks_.erase(found);
}
void Playback::clear_all(EventSink& sink) { while (!tracks_.empty()) clear(tracks_.begin()->first, sink); }
void Playback::set_time(std::uint32_t index, float value) {
    nonnegative(value, "setTrackTime", "timeSeconds"); auto& e = require_track(index, "setTrackTime").chain.back(); e.time = value; e.cursor_initialized = true;
}
void Playback::set_animation_time(float value) {
    nonnegative(value, "setAnimationTime", "timeSeconds"); detached_time_ = value;
    const auto found = tracks_.find(0); if (found == tracks_.end()) return;
    auto& e = found->second.chain.back(); e.time = std::clamp(value, e.start, e.end) - e.start; e.cursor_initialized = true;
}
void Playback::set_options(std::uint32_t index, const TrackOptions& o) {
    constexpr auto* op = "setTrackOptions"; index_valid(index, op);
    unit(o.alpha, op, "alpha"); unit(o.event_threshold, op, "eventThreshold"); unit(o.attachment_threshold, op, "attachmentThreshold"); unit(o.draw_order_threshold, op, "drawOrderThreshold");
    if (o.time_scale) finite(*o.time_scale, op, "timeScale");
    if (o.blend && *o.blend != TrackBlend::replace && *o.blend != TrackBlend::additive) throw Error(ErrorCode::invalid_argument, op, "Unknown track blend.", "blend");
    auto& e = require_track(index, op).chain.back();
    if (o.alpha) e.alpha = *o.alpha;
    if (o.time_scale) e.rate = *o.time_scale;
    if (o.event_threshold) e.event_threshold = *o.event_threshold;
    if (o.attachment_threshold) e.attachment_threshold = *o.attachment_threshold;
    if (o.draw_order_threshold) e.draw_order_threshold = *o.draw_order_threshold;
    if (o.looping) e.looping = *o.looping;
    if (o.hold_previous) e.hold = *o.hold_previous;
    if (o.blend) e.additive = *o.blend == TrackBlend::additive;
}
void Playback::set_range(std::uint32_t index, std::optional<std::pair<float, float>> range) {
    constexpr auto* op = "setTrackAnimationRange"; auto& e = require_track(index, op).chain.back();
    const float start = range ? range->first : 0, end = range ? range->second : e.clip ? e.clip->duration : 0;
    nonnegative(start, op, "start"); nonnegative(end, op, "end");
    if (start > end || end > (e.clip ? e.clip->duration : 0)) throw Error(ErrorCode::invalid_argument, op, "Invalid animation range.", "range");
    e.start = start; e.end = end;
}
void Playback::set_track_end(std::uint32_t index, std::optional<float> value) {
    if (value) nonnegative(*value, "setTrackEnd", "trackEnd");
    require_track(index, "setTrackEnd").chain.back().track_end = value;
}
void Playback::set_mix_duration(std::uint32_t index, float value) {
    nonnegative(value, "setTrackMixDuration", "mixDuration"); require_track(index, "setTrackMixDuration").chain.back().mix_duration = value;
}
void Playback::set_queued_options(std::uint32_t index, std::size_t queue_index, const QueuedEntryOptions& options) {
    constexpr auto* op = "setQueuedEntryOptions";
    if (!options.delay_seconds && !options.mix_duration_seconds && !options.looping) throw Error(ErrorCode::invalid_argument, op, "At least one queued option is required.", "options");
    if (options.delay_seconds) nonnegative(*options.delay_seconds, op, "delaySeconds");
    if (options.mix_duration_seconds) nonnegative(*options.mix_duration_seconds, op, "mixDurationSeconds");
    auto& queue = require_track(index, op).queue;
    if (queue_index >= queue.size()) throw Error(ErrorCode::not_found, op, "Queued entry does not exist.", "queueIndex");
    auto& e = queue[queue_index]; if (options.delay_seconds) e.delay = *options.delay_seconds;
    if (options.mix_duration_seconds) e.mix_duration = *options.mix_duration_seconds;
    if (options.looping) e.looping = *options.looping;
}
void Playback::remove_queued(std::uint32_t index, std::size_t queue_index, EventSink& sink) {
    auto& queue = require_track(index, sink.operation().c_str()).queue;
    if (queue_index >= queue.size()) throw Error(ErrorCode::not_found, sink.operation(), "Queued entry does not exist.", "queueIndex");
    sink.lifecycle(EventKind::dispose, index, queue[queue_index].event_view(), 0);
    queue.erase(queue.begin() + static_cast<std::ptrdiff_t>(queue_index));
}
void Playback::remove_queued_identity(std::uint32_t index, std::uint64_t identity, EventSink& sink) {
    const auto& queue = require_track(index, sink.operation().c_str()).queue;
    for (std::size_t i = 0; i < queue.size(); ++i) if (queue[i].identity == identity) { remove_queued(index, i, sink); return; }
    throw Error(ErrorCode::invalid_state, sink.operation(), "Entry is not queued in this configuration baseline.", "queueIndex");
}
TrackEntry& Playback::find_entry(std::uint32_t index, std::uint64_t identity, const char* op) {
    auto& t = require_track(index, op);
    for (auto& e : t.chain) if (e.identity == identity) return e;
    for (auto& e : t.queue) if (e.identity == identity) return e;
    throw Error(ErrorCode::invalid_state, op, "Entry identity is absent from this configuration baseline.", "trackIndex");
}
std::optional<TrackState> Playback::query(std::uint32_t index) const {
    index_valid(index, "queryTrackState"); const auto found = tracks_.find(index); if (found == tracks_.end()) return {};
    const auto& track = found->second; const auto& e = track.chain.back(); TrackState state;
    state.track_index = index; if (e.clip) state.animation_id = e.clip->id;
    state.track_time = e.time; state.animation_time = e.pose_time(); state.animation_start = e.start; state.animation_end = e.end;
    state.delay_seconds = e.delay; state.track_end = e.track_end; state.alpha = e.alpha; state.time_scale = e.rate;
    state.looping = e.looping; state.hold_previous = e.hold; state.blend = e.additive ? TrackBlend::additive : TrackBlend::replace;
    state.mix_time = e.mix_time; state.mix_duration = e.mix_duration; state.mix_progress = progress(track, track.chain.size() - 1);
    state.event_threshold = e.event_threshold; state.attachment_threshold = e.attachment_threshold; state.draw_order_threshold = e.draw_order_threshold;
    state.queued_entry_count = track.queue.size(); return state;
}
std::vector<QueuedEntryState> Playback::queued(std::uint32_t index) const {
    index_valid(index, "queryQueuedEntries"); std::vector<QueuedEntryState> values;
    const auto found = tracks_.find(index); if (found == tracks_.end()) return values;
    values.reserve(found->second.queue.size());
    for (const auto& e : found->second.queue) values.push_back({index, values.size(), e.clip ? std::optional<std::string>(e.clip->id) : std::nullopt, e.delay, e.mix_duration, e.looping});
    return values;
}
}
